src/default-request-executor.js

```diff
@@ -0,0 +1,122 @@
+/*!
+ * Copyright (c) 2018-present, Okta, Inc. and/or its affiliates. All rights reserved.
+ * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
+ *
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *
+ * See the License for the specific language governing permissions and limitations under the License.
+ */
+
+const RequestExecutor = require('./request-executor');
+const deepCopy = require('deep-copy');
+
+class DefaultRequestExecutor extends RequestExecutor {
+  constructor(config = {}) {
+    super();
+
+    if (config.maxRetries && config.maxRetries < 0) {
+      throw new Error(`okta.client.rateLimit.maxRetries provided as ${config.maxRetries} but must be 0 (disabled) or greater than zero`);
+    }
+
+    if (config.requestTimeout && config.requestTimeout < 0) {
+      throw new Error(`okta.client.rateLimit.requestTimeout provided as ${config.requestTimeout} but must be 0 (disabled) or greater than zero`);
+    }
+
+    this.requestTimeout = config.requestTimeout || 0;
+    this.maxRetries = config.maxRetries === undefined ? 2 : config.maxRetries;
+    this.retryCountHeader = 'X-Okta-Retry-Count';
+    this.retryForHeader = 'X-Okta-Retry-For';
+  }
+
+  buildRetryRequest(request, response) {
+    const newRequest = deepCopy(request);
+    const requestId = this.getOktaRequestId(response);
+    if (!newRequest.headers) {
+      newRequest.headers = {};
+    }
+    if (!newRequest.headers[this.retryForHeader]) {
+      newRequest.headers[this.retryForHeader] = requestId;
+    }
+    newRequest.headers[this.retryCountHeader] =
+      newRequest.headers[this.retryCountHeader] ?
+      newRequest.headers[this.retryCountHeader] + 1 : 1;
+    return newRequest;
+  }
+
+  canRetryRequest(response) {
+    // Validate that we don't have duplicate headers, see OKTA-112507
+    // Duplicate headers are returned by fetch as a comma separated list.
+    const retryHeader = this.getRateLimitReset(response);
+    return !!(retryHeader && !retryHeader.includes(','));
+  }
+
+  fetch(request) {
+    if (request.startTime) {
+      if (this.requestTimeout > 0) {
+        const delta = this.requestTimeout - (new Date() - request.startTime);
```

Comment:
This still feels like we're not timing out at the time the developer would expect. For example:
- Scenario:
  - requestTimeout 30 seconds
  - request that takes 1 second, 429's with an x-rate-limit-reset header for 2 minutes later
- Our code will:
  - Make the initial request with a timeout of 30 seconds, it returns in 1s
  - parseResponse, and set a timeout for 2 minutes later
  - make another request with a 1 ms timeout
  - return with a failed request 2 minutes + 1s + 1ms, rather than 30 seconds

Two problems - we're making an extra request we don't need to that will fail immediately, and we're returning outside of the bounds of the request timeout. Ideally, the logic for this edge case would be that we'd not make that second request and just return immediately because we know that this delta will exceed our timeout. What do you think?

Reply:
Thats a good point, I took out the post-response check that would test if we've passed the request timeout (and not set our own retry setTimeout if so).  That should go back in.  However the check you're also commenting on here (delta < 0 ) I think still needs to be here.  My concern is that the setTimout from the underlying fetch, created some time after `startTime` is stored, will resolve at a time that in the future which causes the delta value to go negative, which should not be passed on to another fetch call because it will mean "no timeout".  I debated between throwing our own error in this case, or retrying the request with 1ms.  I chose the latter so that the error messages would be consistent, but I'm not attached to it.